src/main/java/org/truffleruby/core/array/ArrayAppendOneNode.java

```diff
@@ -60,12 +60,26 @@ public DynamicObject appendOneSameType(DynamicObject array, Object value,
     // Append forcing a generalization
 
     @Specialization(guards = {
-            "strategy.matches(array)", "!strategy.accepts(value)", "valueStrategy.specializesFor(value)",
+            "strategy.isStorageMutable()", "strategy.matches(array)", "!strategy.accepts(value)", "valueStrategy.specializesFor(value)",
     }, limit = "ARRAY_STRATEGIES")
     public DynamicObject appendOneGeneralize(DynamicObject array, Object value,
             @Cached("of(array)") ArrayStrategy strategy,
             @Cached("forValue(value)") ArrayStrategy valueStrategy,
             @Cached("strategy.generalize(valueStrategy)") ArrayStrategy generalizedStrategy) {
+        return appendOneGeneralize(array, value, strategy, generalizedStrategy);
+    }
+
+    @Specialization(guards = {
+            "!strategy.isStorageMutable()", "strategy.matches(array)", "valueStrategy.specializesFor(value)",
+    }, limit = "ARRAY_STRATEGIES")
```

Comment:
Can you order the guards so the common guards with the above specialization are first?